### DISCUSSION TITLE
Connect to Iceberg table on GCS

### DISCUSSION THREAD

**Discussion by vanphuoc3012** [apache/amoro#3135] [Q&A] [answered]
Hi, I try to connect to Iceberg table on GCS and Hive Metastore

I can connect to Hive Metastore and list all schemas and table, but when I access to table, alway failed

I have tried jar file into lib folder:
- add gcs-connector jar version [hadoop3-2.2.11](https://mvnrepository.com/artifact/com.google.cloud.bigdataoss/gcs-connector/hadoop3-2.2.11)

but not work

**Comment by vanphuoc3012** [comment, marked as answer]
I use shaded version and it worked 

https://repo1.maven.org/maven2/com/google/cloud/bigdataoss/gcs-connector/hadoop3-2.2.10/gcs-connector-hadoop3-2.2.10-shaded.jar

**Comment by krystian-blaszczyk** [comment]
I am experiencing several issues while attempting to connect Amoro AMS to an external Iceberg REST catalog (Apache Polaris) using Google Cloud Storage (GCS) as the storage layer and Polaris vended-credentials for authentication.

Despite adding the necessary JAR files to the lib directory (including gcs-connector-hadoop3-shaded and iceberg-gcp-bundle), I have encountered the following roadblocks:

1. Classloader / ClassNotFoundException: Amoro AMS reports java.lang.ClassNotFoundException: org.apache.iceberg.gcp.gcs.GCSFileIO, even when the JAR files are present in /usr/local/amoro/lib. In a containerized environment, the AMS process seems to look for libraries only in specific subdirectories (like /ams/lib). Forcing io-impl in the catalog configuration prevents the entire AMS dashboard from starting (resulting in an "AMS start error").

2. Fallback to HadoopFileIO Failure: When the io-impl parameter is not explicitly set, Amoro attempts to use HadoopFileIO by default. This leads to the error: org.apache.hadoop.fs.UnsupportedFileSystemException: No FileSystem for scheme "gs". This indicates that the native GCS support from the iceberg-gcp-bundle is not being activated automatically for gs:// paths.
 
3. Metadata Server Timeout with vended-credentials: When using Polaris, the REST server should delegate temporary credentials. However, the Amoro logs show that the GCS connector ignores the credentials provided by Polaris and instead attempts to connect to the Google Cloud metadata server (http://169.254.169.254/...), which results in a SocketTimeoutException.

Simply adding the gcs-connector-shaded library (as suggested in issue #3135) does not resolve the issue in a Polaris + Amoro architecture. The system struggles to correctly load GCP classes into the Classpath and fails to prioritize GCSFileIO over HadoopFileIO when handling delegated access tokens.

**Comment by vanphuoc3012** [comment]
when create catalog, in hive-site.xml, add gs implement config. Here is my hive-site.xml:

```xml
<?xml version="1.0" encoding="UTF-8"?>
 <configuration>
    <property>
        <name>fs.gs.impl</name>
        <value>com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem</value>
    </property>
    <property>
        <name>fs.AbstractFileSystem.gs.impl</name>
        <value>com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS</value>
    </property>

    <property>
        <name>fs.gs.status.parallel.enable</name>
        <value>false</value>
    </property>

    <property>
        <name>fs.gs.max.requests.per.batch</name>
        <value>30</value>
    </property>

    <property>
        <name>fs.gs.batch.threads</name>
        <value>30</value>
    </property>

    <property>
        <name>fs.gs.list.max.items.per.call</name>
        <value>50000</value>
    </property>

    
    <property>
        <name>fs.gs.implicit.dir.repair.enable</name>
        <value>false</value>
    </property>
    </configuration>
```

**Reply by krystian-blaszczyk** [reply]
I use Iceberg REST Catalog not Hive/Hadoop
